ocaml/xapi/xapi_mgmt_iface.ml

```diff
@@ -48,114 +30,151 @@ let update_mh_info interface =
+    info "Restarting stunnel (accepting connections on %s)" accept ;
+    (* cache `accept` so client can call `reconfigure_stunnel` easily *)
+    Mutex.execute stunnel_accept_m (fun () -> stunnel_accept := Some accept) ;
+    _restart_stunnel_no_cache ~__context ~accept
+
+  let reconfigure_stunnel ~__context =
+    let f =
+      Mutex.execute stunnel_accept_m (fun () ->
+          match !stunnel_accept with
+          | None ->
+              fun () ->
+                D.warn
+                  "reconfigure_stunnel: accept is not set, so not restarting \
+                   stunnel"
+          | Some accept ->
+              fun () -> _restart_stunnel_no_cache ~__context ~accept)
+    in
+    f ()
+end
+
+module Server = struct
+  (* Keep track of the management interface server thread.
+     Stores a key into the table in Http_srv which identifies the server thread bound
+     to the management IP. *)
+  let management_servers = ref []
+
+  let stop () =
+    debug "Shutting down the old management interface (if any)" ;
+    List.iter (fun i -> Http_svr.stop i) !management_servers ;
+    management_servers := []
+
+  (* Even if xapi listens on all IP addresses, there is still an interface appointed as
+     _the_ management interface. Hosts in a pool use the IP addresses of this interfaces
+     to communicate with each other. *)
+  let start ~__context ?addr () =
+    let socket, accept =
+      match addr with
+      | None -> (
+          info "Starting new server (listening on all IP addresses)" ;
+          try
+            (* Is it IPv6 ? *)
+            let addr = Unix.inet6_addr_any in
+            ( Xapi_http.bind (Unix.ADDR_INET (addr, Constants.http_port))
+            , ":::443" )
+          with _ ->
+            (* No. *)
+            let addr = Unix.inet_addr_any in
+            (Xapi_http.bind (Unix.ADDR_INET (addr, Constants.http_port)), "443")
+        )
+      | Some ip -> (
+          info "Starting new server (listening on %s)" ip ;
+          let addr = Unix.inet_addr_of_string ip in
+          let sockaddr = Unix.ADDR_INET (addr, Constants.http_port) in
+          ( Xapi_http.bind sockaddr
+          , match Unix.domain_of_sockaddr sockaddr with
+            | Unix.PF_INET6 ->
+                "::1:443"
+            | _ ->
+                "127.0.0.1:443" )
+        )
+    in
+    Http_svr.start Xapi_http.server socket ;
+    management_servers := socket :: !management_servers ;
+    Stunnel.restart_stunnel ~__context ~accept ;
+    if Pool_role.is_master () && addr = None then
+      (* NB if we synchronously bring up the management interface on a master with a blank
+         database this can fail... this is ok because the database will be synchronised later *)
+      Server_helpers.exec_with_new_task "refreshing consoles" (fun __context ->
+          Dbsync_master.set_master_ip ~__context ;
+          Dbsync_master.refresh_console_urls ~__context)
+
+  type listening_mode = Off | Any | Local of StringSet.t
```

Comment:
StringSet might not be the best name - it's a set of addresses (which are strings).